File: leetcode/merge-k-sorted-lists/merge-k-sorted-lists-pq.py
# LeetCode #23 - Merge k Sorted Lists
# https://leetcode.com/problems/merge-k-sorted-lists/
# Using a binary minheap priority queue.

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    @staticmethod
    def mergeKLists(lists: List[ListNode]) -> ListNode:
        sentinel = ListNode()
        pre = sentinel

        heap = [Box(node) for node in lists if node]
        heapq.heapify(heap)

        while heap:
            node = heapq.heappop(heap).node
            pre.next = node
            pre = pre.next
            if node.next:
                heapq.heappush(heap, Box(node.next))

        pre.next = None
        return sentinel.next


@functools.total_ordering
class Box:
    def __init__(self, node):
        self.node = node

    def __eq__(self, other):
        return self.node.val == other.node.val

    def __lt__(self, other):
        return self.node.val < other.node.val

File: leetcode/merge-k-sorted-lists/merge-k-sorted-lists-pq.java
// LeetCode #23 - Merge k Sorted Lists
// https://leetcode.com/problems/merge-k-sorted-lists/
// Using a priority queue.

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        var sentinel = new ListNode();
        var pre = sentinel;

        for (var heap = getInitialHeap(lists); !heap.isEmpty(); ) {
            var node = heap.remove();
            pre.next = node;
            pre = pre.next;
            if (node.next != null) heap.add(node.next);
        }

        pre.next = null;
        return sentinel.next;
    }

    private static Queue<ListNode> getInitialHeap(ListNode[] lists) {
        Queue<ListNode> heap =
            new PriorityQueue<>(Comparator.comparing(node -> node.val));

        Stream.of(lists).filter(Objects::nonNull).forEachOrdered(heap::add);

        return heap;
    }
}
